import pandas as pd 
import yaml

sample_names = list(pd.read_csv(config["demux"]["sample_sheet"])["sample_name"])
print(sample_names)

rule all:
    input:
        "outputs/gene_counts.tsv",
        "outputs/aligned/align_summary.csv"

rule get_fasta_files:
    output: 
        r1="res/r1.fastq.gz",
        r2="res/r2.fastq.gz"
    params:
        r1_loc=config["demux"]["r1"],
        r2_loc=config["demux"]["r2"]
    shell:
        """
        echo "bulk_step,get_data,running" >> logs/bulk_progress.log
        wget -O res/r1.fastq.gz {params.r1_loc}
        wget -O res/r2.fastq.gz {params.r2_loc}
        echo "bulk_step,get_data,finished" >> logs/bulk_progress.log
        """

rule idemux:
    input:
        r1="res/r1.fastq.gz",
        r2="res/r2.fastq.gz",
        sample_sheet=config["demux"]["sample_sheet"]
    output:
        out_dir=directory("outputs/idemultiplexed"),
        r1_fastq=expand("outputs/idemultiplexed/{sample}_R1.fastq.gz", sample=sample_names),
        r2_fastq=expand("outputs/idemultiplexed/{sample}_R2.fastq.gz", sample=sample_names)
    shell:
        """
        echo "bulk_step,demux,running" >> logs/bulk_progress.log
        idemux --r1 {input.r1} --r2 {input.r2} --sample-sheet {input.sample_sheet} --out {output.out_dir}
        echo "bulk_step,demux,finished" >> logs/bulk_progress.log
        """

rule extract_umi:
    input:
        r1= "outputs/idemultiplexed/{sample}_R1.fastq.gz",
        r2= "outputs/idemultiplexed/{sample}_R2.fastq.gz"
    params:
        extract_method="string",
        bc_pattern="X",
        bc_pattern2="NNNNNNNNNN"
    log: "logs/{sample}_extraction_log"
    output:
        r1_extracted="outputs/extracted/{sample}_extracted_R1.fastq.gz"
    shell:
        """
        echo "{wildcards.sample},extract,running" >> logs/progress.log
        umi_tools extract --extract-method={params.extract_method} --bc-pattern {params.bc_pattern} \
        --bc-pattern2 {params.bc_pattern2} -L {log} -S {output.r1_extracted} \
        -I {input.r1} --read2-in={input.r2} --read2-out=/dev/null
        echo "{wildcards.sample},extract,finished" >> logs/progress.log
        """

rule cutadapt:
    input:
        r1_extracted="outputs/extracted/{sample}_extracted_R1.fastq.gz"
    output:
        r1_trimmed="outputs/trimmed/{sample}_extracted_trimmed_R1.fastq.gz"
    threads: 4
    shell:
        """
        echo "{wildcards.sample},trim,running" >> logs/progress.log
        cutadapt --quiet -m 20 -O 20 -a "polyA=A{{20}}" -a "QUALITY=G{{20}}" -n {threads} {input.r1_extracted} | \
        cutadapt --quiet -m 20 -O 3 --nextseq-trim=10 -a "r1adapter=AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC;min_overlap=3;max_error_rate=0.100000" - | \
        cutadapt --quiet -m 20 -O 3 -a "r1polyA=A{{18}}" - | \
        cutadapt --quiet -m 20 -O 20 -g "r1adapter=AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC;min_overlap=20" --discard-trimmed -o {output.r1_trimmed} -
        echo "{wildcards.sample},trim,finished" >> logs/progress.log
        """

rule get_genome_and_annotation:
    output:
        genome_fasta="res/genome/genome_fasta.fastq",
        genome_annotation="res/genome/genome_annotation.gtf"
    params:
        genome_fasta_location=config["align"]["genome_fasta_location"],
        genome_annotation_location=config["align"]["genome_annotation_location"]
    threads: 1
    shell:
        """
        mkdir -p res/genome
        wget -O res/genome/genome_fasta.fastq.gz {params.genome_fasta_location}
        wget -O res/genome/genome_annotation.gtf.gz {params.genome_annotation_location}
        gunzip res/genome/genome_fasta.fastq.gz
        gunzip res/genome/genome_annotation.gtf.gz
        """

rule make_star_index:
    input:
        genome_fasta="res/genome/genome_fasta.fastq",
        genome_annotation="res/genome/genome_annotation.gtf"
    output: 
        "res/genome/Genome",
        "res/genome/Log.out"
    threads: 10
    params:
        sjdb_overhang=config["align"]["sjdb_overhang"]
    shell:
        """
        echo "bulk_step,make_index,running" >> logs/bulk_progress.log
        STAR --runMode genomeGenerate --runThreadN {threads} \
        --genomeDir res/genome --genomeFastaFiles {input.genome_fasta} \
        --sjdbGTFfile {input.genome_annotation} --sjdbOverhang {params.sjdb_overhang}
        echo "bulk_step,make_index,finished" >> logs/bulk_progress.log
        """

rule star_align:
    input:
        r1_trimmed="outputs/trimmed/{sample}_extracted_trimmed_R1.fastq.gz",
        star_was_indexed="res/genome/Genome"
    params:
        prefix="outputs/aligned/{sample}.",
        index_loc="res/genome"
    log: "outputs/aligned/{sample}.Log.final.out"
    output: 
        "outputs/aligned/{sample}.Aligned.sortedByCoord.out.bam",
    threads: config["align"]["threads"]
    shell:
        """
        echo "{wildcards.sample},align,running" >> logs/progress.log
        STAR --runThreadN {threads} --readFilesCommand zcat --genomeDir {params.index_loc} --readFilesIn {input.r1_trimmed} \
        --outFilterType BySJout --outFilterMultimapNmax 200 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 \
        --outFilterMismatchNoverLmax 0.6 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --limitOutSJcollapsed 5000000 \
        --limitIObufferSize 500000 500000 --outSAMattributes NH HI NM MD --outSAMtype BAM SortedByCoordinate --outFileNamePrefix {params.prefix} \
        --limitBAMsortRAM 500000 --genomeLoad LoadAndKeep
        echo "{wildcards.sample},align,finished" >> logs/progress.log
        """

rule align_log_summary:
    input: expand("outputs/aligned/{sample}.Log.final.out", sample=sample_names)
    output: "outputs/aligned/align_summary.csv"
    params: str_samples=" ".join(sample_names)
    shell: 
        """
        python src/scripts/parse_logs.py {params.str_samples}
        """



rule index:
    input: "outputs/aligned/{sample}.Aligned.sortedByCoord.out.bam"
    output: "outputs/aligned/{sample}.Aligned.sortedByCoord.out.bam.bai"
    shell:
        """
        samtools index {input}
        """

rule dedup:
    input: 
        alignment="outputs/aligned/{sample}.Aligned.sortedByCoord.out.bam",
        index="outputs/aligned/{sample}.Aligned.sortedByCoord.out.bam.bai"
    output: "outputs/dedup/{sample}.dedup.bam"
    shell:
        """
        echo "{wildcards.sample},dedup,running" >> logs/progress.log
        umi_tools dedup -I {input.alignment} -S {output} --multimapping-detection-method=NH \
        --output-stats=outputs/dedup/deduplicated.txt --log=outputs/dedup/deduplication.log
        echo "{wildcards.sample},dedup,finished" >> logs/progress.log
            """
rule index2:
    input: "outputs/dedup/{sample}.dedup.bam"
    output: "outputs/dedup/{sample}.dedup.bam.bai"
    shell:
        """
        samtools index {input}
        """
rule featureCounts:
    input: 
        alignments=expand("outputs/dedup/{sample}.dedup.bam", sample=sample_names)
    params:
        annotation="res/genome/genome_annotation.gtf"
    output: "outputs/gene_counts.tsv"
    shell:
        """
        echo "bulk_step,feature_counts,running" >> logs/bulk_progress.log
        featureCounts -s 1 -T 4 -t exon -g gene_id -a {params.annotation} \
        -o {output} {input.alignments}
        echo "bulk_step,feature_counts,finished" >> logs/bulk_progress.log
        """